src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -168,61 +168,64 @@ public IODataResourceTypeContext TypeContext
         /// <param name="serializationInfo">The serialization info of the resource for writing without model.</param>
         /// <param name="actualEntityType">The edm entity type of the resource</param>
         /// <param name="requiresId">Whether key properties are required to be returned</param>
-        /// <returns>Key value pair array</returns>
-        internal static KeyValuePair<string, object>[] GetKeyProperties(
+        /// <returns>Key value pair list.</returns>
+        internal static IList<KeyValuePair<string, object>> GetKeyProperties(
             ODataResourceBase resource,
             ODataResourceSerializationInfo serializationInfo,
             IEdmEntityType actualEntityType,
             bool requiresId)
         {
             Debug.Assert(resource != null, "GetKeyProperties called for a null resource.");
 
-            KeyValuePair<string, object>[] keyProperties = null;
-            string actualEntityTypeName = string.IsNullOrEmpty(resource.TypeName) ? actualEntityType?.FullName() : resource.TypeName;
+            IList<KeyValuePair<string, object>> keyProperties = null;
 
             // if we have serializationInfo, try that first
             if (serializationInfo != null)
             {
-                keyProperties = ODataResourceMetadataContextWithoutModel.GetPropertiesBySerializationInfoPropertyKind(resource, ODataPropertyKind.Key, actualEntityTypeName);
+                keyProperties = ODataResourceMetadataContextWithoutModel.GetPropertiesBySerializationInfoPropertyKind(resource, ODataPropertyKind.Key, actualEntityType);
             }
 
             // if we didn't get any keys from serializationInfo, try using entity type
-            if ((keyProperties == null || keyProperties.Length == 0) && actualEntityType != null)
+            if ((keyProperties == null || keyProperties.Count == 0) && actualEntityType != null)
             {
-                keyProperties = GetPropertyValues(actualEntityType.Key(), resource, actualEntityType, requiresId).ToArray();
+                keyProperties = GetPropertyValues(actualEntityType.Key(), resource, actualEntityType, requiresId);
             }
 
-            if (!ValidateEntityTypeHasKeyProperties(keyProperties, actualEntityTypeName, requiresId))
-            { 
+            if (!ValidateEntityTypeHasKeyProperties(keyProperties, resource, actualEntityType, requiresId))
+            {
                 return Enumerable.Empty<KeyValuePair<string, object>>().ToArray();
             }
 
             return keyProperties;
         }
 
-        private static IEnumerable<KeyValuePair<string, object>> GetPropertyValues(IEnumerable<IEdmStructuralProperty> properties, ODataResourceBase resource, IEdmEntityType actualEntityType, bool isRequired)
+        private static List<KeyValuePair<string, object>> GetPropertyValues(IEnumerable<IEdmStructuralProperty> properties, ODataResourceBase resource, IEdmStructuredType actualEntityType, bool isRequired)

```

Comment:
Follow up question
Is the method called multiple times?

Reply:
This method is used to retrieve the values of key properties and ETag properties. So, it would be called, for example, whenever we want to build the key segment, which we do for every entity resource in the response (whether top-level or navigation property).